### PR TITLE
Initial hw2 for Mikihisa Yuasa (myuasa2)

### PR DESCRIPTION
**Pull request by miki-yuasa**

### CODE REVIEW

**Comment by DavidMSCode** [comment]
Hi Mikihisa,
I'll review your code today

File: dqn/dqn.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+        super().__init__()
+        self.layer1 = nn.Linear(n_observations, 64)
+        self.layer2 = nn.Linear(64, 64)
+        self.layer3 = nn.Linear(64, n_actions)
+
+    def forward(self, x: Tensor) -> Tensor:
+        x = nn.functional.tanh(self.layer1(x))
+        x = nn.functional.tanh(self.layer2(x))
+        x = self.layer3(x)
+        return x
```
Comment (DavidMSCode) [inline dqn/dqn.py:16]: DQN class looks good

File: dqn/replay.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        self.memory: deque[Transition] = deque([], maxlen=capacity)
+
+    def push(self, transition: Transition) -> None:
+        self.memory.append(transition)
+
+    def sample(self, batch_size: int) -> list[Transition]:
+        return random.sample(self.memory, batch_size)
+
+    def __len__(self) -> int:
+        return len(self.memory)
```
Comment (DavidMSCode) [inline dqn/replay.py:26]: Memory class looks good

File: dqn/utils.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        "t": [0],
+        "s": [s],
+        "a": [],
+        "r": [],
+    }
+
+    # Simulate until episode is done
+    done = False
+    while not done:
+        a = random.randrange(env.num_actions)
```
Comment (DavidMSCode) [inline dqn/utils.py]: Action should be chosen from the learned policy

File: train_discreteaction_pendulum.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        state = next_state
+
+        # Perform one step of the optimization (on the policy network)
+        optimize_model()
+
+        # Soft update of the target network's weights
+        # θ′ ← τ θ + (1 −τ )θ′
+        target_net_state_dict = target_net.state_dict()
+        policy_net_state_dict = policy_net.state_dict()
+        for key in policy_net_state_dict:
```
Comment (DavidMSCode) [inline train_discreteaction_pendulum.py]: The algo from the paper doesn't use a soft update but resets the target network to the policy network after C steps. This does still work though.

File: dqn/utils.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+import numpy as np
+from numpy import ndarray
+import torch
+
+from discreteaction_pendulum import Pendulum
+from dqn.dqn import DQN
+
+
+class PolicyWrapper:
```
Comment (DavidMSCode) [inline dqn/utils.py]: I like the wrapper class for inputting into the video function

File: dqn/utils.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    ax[0].plot(data["t"], theta, label="theta")
+    ax[0].plot(data["t"], thetadot, label="thetadot")
+    ax[0].legend()
+    ax[1].plot(data["t"][:-1], tau, label="tau")
+    ax[1].legend()
+    ax[2].plot(data["t"][:-1], data["r"], label="r")
+    ax[2].legend()
+    ax[2].set_xlabel("time step")
+    plt.tight_layout()
+    plt.savefig(traj_filename)
```
Comment (DavidMSCode) [inline dqn/utils.py]: besides the random actions the plotting funcs look good

File: train_discreteaction_pendulum.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+import torch.nn as nn
+import torch.optim as optim
+
+from discreteaction_pendulum import Pendulum
+from dqn.dqn import DQN
+from dqn.replay import ReplayMemory, Transition
+from dqn.utils import generate_trajectory
+
+
+BATCH_SIZE = 128
```
Comment (DavidMSCode) [inline train_discreteaction_pendulum.py:20]: Hyperparameters look reasonable

File: train_discreteaction_pendulum.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+
+env = Pendulum()
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+# Get number of actions from gym action space
+n_actions = env.num_actions
+# Get the number of state observations
+state = env.reset()
+n_observations = len(state)
+
+policy_net = DQN(n_observations, n_actions).to(device)
```
Comment (DavidMSCode) [inline train_discreteaction_pendulum.py]: policy and target net init look good

File: train_discreteaction_pendulum.py [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+memory = ReplayMemory(10000)
+
+
+steps_done = 0
+
+
+def select_action(state) -> torch.Tensor:
+    global steps_done
+    sample = random.random()
+    eps_threshold = EPS_END + (EPS_START - EPS_END) * math.exp(
```
Comment (DavidMSCode) [inline train_discreteaction_pendulum.py]: epsilon decay looks good

File: train_discreteaction_pendulum.py [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+
+        if done:
+            episode_durations.append(t + 1)
+            break
+
+print("Complete")
+
+traj_filename: str = "figures/traj_w_rep_w_target.png"
+video_filename: str = "figures/video_w_rep_w_target.gif"
+generate_trajectory(env, policy_net, traj_filename, video_filename, device)
```
Comment (DavidMSCode) [inline train_discreteaction_pendulum.py]: Missing learning curve, policy plot, state-value plot and ablation study.

File: train_discreteaction_pendulum.py [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+
+    # Optimize the model
+    optimizer.zero_grad()
+    loss.backward()
+    # In-place gradient clipping
+    torch.nn.utils.clip_grad_value_(policy_net.parameters(), 100)
+    optimizer.step()
+
+
+if torch.cuda.is_available():
```
Comment (DavidMSCode) [inline train_discreteaction_pendulum.py]: I would expect torch with CUDA to run faster than on the CPU. Probably need another order of magnitude or two for a decent agent. Though I'm guessing the low values were just for quick testing.

File: train_discreteaction_pendulum.py [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+    criterion = nn.SmoothL1Loss()
+    loss: torch.Tensor = criterion(
+        state_action_values, expected_state_action_values.unsqueeze(1)
+    )
+
+    # Optimize the model
+    optimizer.zero_grad()
+    loss.backward()
+    # In-place gradient clipping
+    torch.nn.utils.clip_grad_value_(policy_net.parameters(), 100)
```
Comment (DavidMSCode) [inline train_discreteaction_pendulum.py]: Apparently gradient clipping should happen between zero_grad() and the backward() propagation.

File: train_discreteaction_pendulum.py [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+    # Compute V(s_{t+1}) for all next states.
+    # Expected values of actions for non_final_next_states are computed based
+    # on the "older" target_net; selecting their best reward with max(1)[0].
+    # This is merged based on the mask, such that we'll have either the expected
+    # state value or 0 in case the state was final.
+    next_state_values = torch.zeros(BATCH_SIZE, device=device)
+    with torch.no_grad():
+        next_state_values[non_final_mask] = target_net(non_final_next_states).max(1)[0]
+    # Compute the expected Q values
+    expected_state_action_values = (next_state_values * GAMMA) + reward_batch
```
Comment (DavidMSCode) [inline train_discreteaction_pendulum.py]: state value update looks good

File: train_discreteaction_pendulum.py [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+        plt.plot(means.numpy())
+
+    plt.pause(0.001)  # pause a bit so that plots are updated
+
+
+def optimize_model():
+    if len(memory) < BATCH_SIZE:
+        return
+    transitions = memory.sample(BATCH_SIZE)
+    # Transpose the batch (see https://stackoverflow.com/a/19343/3343043 for
```
Comment (DavidMSCode) [inline train_discreteaction_pendulum.py]: Sampling memory looks good

File: train_discreteaction_pendulum.py [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+    torch.nn.utils.clip_grad_value_(policy_net.parameters(), 100)
+    optimizer.step()
+
+
+if torch.cuda.is_available():
+    num_episodes = 10
+else:
+    num_episodes = 50
+
+for i_episode in range(num_episodes):
```
Comment (DavidMSCode) [inline train_discreteaction_pendulum.py]: episode sim looks good

**Review by DavidMSCode — COMMENTED** [review]
Algorithm looks good, just missing the plots really

File: dqn/utils.py (hunk 3 above)
Comment (miki-yuasa) [reply dqn/utils.py]: fixed

File: train_discreteaction_pendulum.py (hunk 10 above)
Comment (miki-yuasa) [reply train_discreteaction_pendulum.py]: added
